broker/src/test/java/io/camunda/zeebe/broker/exporter/stream/ExternalExporterContainerTest.java

```diff
@@ -0,0 +1,203 @@
+        .isSameAs(expectedClassLoader)
+        .isInstanceOf(ExporterJarClassLoader.class);
+  }
+
+  @Test
+  void shouldSetTclOnExport(final @TempDir File jarDirectory)
+      throws ExporterLoadException, IOException {
+    // given
+    final var exporterClass = createUnloadedExporter();
+    final var jarFile = createExporterJar(new File(jarDirectory, "exporter.jar"), exporterClass);
+    final var descriptor = runtime.loadExternalExporter(jarFile, EXPORTER_CLASS_NAME);
+    final var expectedClassLoader = descriptor.newInstance().getClass().getClassLoader();
+    final var container = runtime.newContainer(descriptor);
+
+    // when
+    final var record = mock(TypedRecord.class);
+    // set a high position to ensure we export it
+    when(record.getPosition()).thenReturn(Long.MAX_VALUE);
+    container.exportRecord(mock(RecordMetadata.class), record);
+
+    // then
+    final var exporterInstance = (TclExporter) container.getExporter();
+    assertThat(exporterInstance.onExportTCL)
+        .isSameAs(expectedClassLoader)
+        .isInstanceOf(ExporterJarClassLoader.class);
+  }
+
+  @Test
+  void shouldSetTclOnClose(final @TempDir File jarDirectory)
+      throws ExporterLoadException, IOException {
+    // given
+    final var exporterClass = createUnloadedExporter();
+    final var jarFile = createExporterJar(new File(jarDirectory, "exporter.jar"), exporterClass);
+    final var descriptor = runtime.loadExternalExporter(jarFile, EXPORTER_CLASS_NAME);
+    final var expectedClassLoader = descriptor.newInstance().getClass().getClassLoader();
+    final var container = new ExporterContainer(descriptor);
+
+    // when
+    container.close();
+
+    // then
+    final var exporterInstance = (TclExporter) container.getExporter();
+    assertThat(exporterInstance.onCloseTCL)
+        .isSameAs(expectedClassLoader)
+        .isInstanceOf(ExporterJarClassLoader.class);
+  }
+
+  private Unloaded<TclExporter> createUnloadedExporter() {
+    return new ByteBuddy().subclass(TclExporter.class).name(EXPORTER_CLASS_NAME).make();
+  }
+
+  private File createExporterJar(final File dest, final Unloaded<TclExporter> unloadedExporter)
+      throws IOException {
+    final var jarFile = unloadedExporter.toJar(dest);
+    final var unloadedConfig =
+        new ByteBuddy().subclass(Object.class).name("com.acme.Config").make();
+
+    unloadedConfig.inject(jarFile);
+    return jarFile;
+  }
+
+  // the class must be visible to the generated exporter for ByteBuddy to delegate method invocation
+  public abstract static class TclExporter implements Exporter {
+    public ClassLoader onConfigureTCL;
+    public ClassLoader onOpenTCL;
+    public ClassLoader onCloseTCL;
+    public ClassLoader onExportTCL;
+
+    @Override
+    public void configure(final Context context) throws Exception {
+      onConfigureTCL = Thread.currentThread().getContextClassLoader();
+
+      // this will cause it to fail when using the wrong class loader during validation
```

Comment:
please also explain why this would fail (that might be lost to future readers)

Reply:
Good point, it seemed obvious to me at the time why it fails, but yeah :sweat_smile: